opendevin/runtime/client/client.py

```
@@ -60,19 +61,77 @@ class RuntimeClient:
+            await plugin.initialize(self.username)
             self.plugins[plugin.name] = plugin
             logger.info(f'Initializing plugin: {plugin.name}')
 
-    def _init_bash_shell(self, work_dir: str) -> None:
-        self.shell = pexpect.spawn('/bin/bash', encoding='utf-8', echo=False)
+        # This is a temporary workaround
+        # TODO: refactor AgentSkills to be part of JupyterPlugin
+        # AFTER ServerRuntime is deprecated
+        if 'agent_skills' in self.plugins and 'jupyter' in self.plugins:
+            obs = await self.run_ipython(
+                IPythonRunCellAction(
+                    code=(
+                        'import sys\n'
+                        'sys.path.insert(0, "/opendevin/code")\n'
+                        # TODO: figure out how to pass added env var to IPython?
+                        'from opendevin.runtime.plugins.agent_skills.agentskills import *'
+                    )
+                )
+            )
+            logger.info(f'AgentSkills initialized: {obs}')
+
+    def _init_user(self, username: str, user_id: int) -> None:
+        """Create user if not exists."""
+        # Skip root since it is already created
+        if username == 'root':
+            return
+
+        # Add sudoer
+        sudoer_line = r"echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers"
+        output = subprocess.run(sudoer_line, shell=True, capture_output=True)
+        if output.returncode != 0:
+            raise RuntimeError(f'Failed to add sudoer: {output.stderr.decode()}')
+        logger.debug(f'Added sudoer successfully. Output: [{output.stdout.decode()}]')
+
+        # Add user
+        output = subprocess.run(
+            (
+                f'useradd -rm -d /home/{username} -s /bin/bash '
+                f'-g root -G sudo -g root -G sudo -u {user_id} {username}'
+            ),
+            shell=True,
+            capture_output=True,
+        )
+        if output.returncode != 0:
+            raise RuntimeError(
```

If you raise here, in the middle of initialization, does this trigger a session restart on the server?
(I'll probably find out reading further down 😄 but it's a lot of code...)

---

Yep.. But I think it'd better to raise an error for us to fix it rather than pretend nothing happens and let the error hide.

---

(and it could be transient issue that can be fixed by restart the server though)

---

Wasn't suggesting to hide the error, just what the consequences were. :)